File: Java/labs/lab13_Proxy/task_2_1 – solution/src/ProjectProxy.java
public class ProjectProxy implements Project {

    private Project project;
    final private String url;

    public ProjectProxy(String url) {
        this.url = url;
    }

    @Override
    public void run() {
        if (project == null){
            project = new RealProject(url);
        }
        project.run();
    }
}

File: Cpp/lab13_Proxy/task_2_1 - solution/src/project_proxy.cpp
#include "project_proxy.hpp"
#include "real_project.hpp"


ProjectProxy::ProjectProxy(std::string const& url)
: url(url) {}

void ProjectProxy::run()
{
  if (this->project == nullptr)
  {
    this->project = new RealProject(this->url);
  }
  project->run();
}
